server/re/src/msi_genquery2.cpp

```diff
@@ -0,0 +1,315 @@
+
+    if (!inserted) {
+        log_msi::error("{}: Could not capture GenQuery2 results.", __func__);
+        return SYS_LIBRARY_ERROR;
+    }
+
+    // Return the handle to the caller.
+    fillStrInMsParam(_handle, std::to_string(key).c_str());
+
+    return 0;
+} // msi_genquery2_execute
+
+/// Moves the cursor forward by one row.
+///
+/// \warning This microservice is experimental and may change in the future.
+/// \warning This microservice is NOT thread-safe.
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// \param[in] _handle The GenQuery2 handle.
+/// \param[in] _rei    This parameter is special and should be ignored.
+///
+/// \return An integer.
+/// \retval       0 If data can be read from the new row.
+/// \retval -408000 If the end of the resultset has been reached.
+/// \retval      <0 If an error occurred.
+///
+/// \b Example
+/// \code{.py}
+/// # A global constant representing the value which indicates the end of
+/// # the resultset has been reached.
+/// END_OF_RESULTSET = -408000;
+///
+/// iterating_over_genquery2_results()
+/// {
+///     # Execute a query. The results are stored in the Rule Engine Plugin.
+///     msi_genquery2_execute(*handle, "select COLL_NAME, DATA_NAME order by DATA_NAME desc limit 1");
+///
+///     # Iterate over the results.
+///     while (true) {
+///         *ec = errorcode(msi_genquery2_next_row(*handle));
+///
+///         if (*ec < 0) {
+///             if (END_OF_RESULTSET == *ec) {
+///                 break;
+///             }
+///
+///             failmsg(*ec, "Unexpected error while iterating over GenQuery2 resultset.");
+///         }
+///
+///         msi_genquery2_column(*handle, '0', *coll_name); # Copy the COLL_NAME into *coll_name.
+///         msi_genquery2_column(*handle, '1', *data_name); # Copy the DATA_NAME into *data_name.
+///         writeLine("stdout", "logical path => [*coll_name/*data_name]");
+///     }
+///
+///     # Free any resources used. This is handled for you when the agent is shut down as well.
+///     msi_genquery2_free(*handle);
+/// }
+/// \endcode
+///
+/// \since 4.3.2
+auto msi_genquery2_next_row(MsParam* _handle, RuleExecInfo* _rei) -> int
+{
+    log_msi::trace(__func__);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->type);
+
+    IRODS_MSI_REQUIRE_TYPE(_handle->type, STR_MS_T);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->inOutStruct);
+
+    try {
+        const auto ctx_handle_index = std::stoll(static_cast<char*>(_handle->inOutStruct));
+
+        if (ctx_handle_index < 0) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        auto iter = gq2_context.find(ctx_handle_index);
+
+        if (std::end(gq2_context) == iter) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+        }
```

Comment:
Should we return an error here? Or is it intentionally let go because the `std::exception` is caught below?

Reply:
Yes. Good catch.

Reply:
Done